pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,288 @@
+	engine := &engine{
+		conn:    conn,
+		uuid:    uuid,
+		metrics: make(map[column]prometheus.Metric),
+	}
+	if err := engine.writeMetadata(); err != nil {
+		return nil, fmt.Errorf("writing metadata: %w", err)
+	}
+	return engine, nil
+}
+
+// writeMetadata writes Promscale and Tobs metadata. Must be written only by
+func (t *engine) writeMetadata() error {
+	promscale, err := promscaleMetadata()
+	if err != nil {
+		return fmt.Errorf("promscale metadata: %w", err)
+	}
+	promscale["promscale_exec_platform"] = ExecPlatform
+	if err := syncTimescaleMetadataTable(t.conn, Stats(promscale)); err != nil {
+		return fmt.Errorf("writing metadata for promscale: %w", err)
+	}
+
+	tobs := tobsMetadata()
+	if len(tobs) > 0 {
+		if err := syncTimescaleMetadataTable(t.conn, Stats(tobs)); err != nil {
+			return fmt.Errorf("writing metadata for tobs: %w", err)
+		}
+	}
+	return nil
+}
+
+// syncTimescaleMetadataTable syncs the metadata/stats with telemetry metadata table and returns the last error if any.
+func syncTimescaleMetadataTable(conn pgxconn.PgxConn, m Stats) error {
+	batch := conn.NewBatch()
+	for key, metadata := range m {
+		safe := pgutf8str.Text{}
+		if err := safe.Set(metadata); err != nil {
+			return fmt.Errorf("setting in pgutf8 safe string: %w", err)
+		}
+		query := "INSERT INTO _timescaledb_catalog.metadata VALUES ( $1, $2, true ) ON CONFLICT (key) DO UPDATE SET value = $2, include_in_telemetry = true WHERE metadata.key = $1"
+		batch.Queue(query, key, safe)
+	}
+	results, err := conn.SendBatch(context.Background(), batch)
+	if err != nil {
+		return fmt.Errorf("error sending batch: %w", err)
+	}
+	rows, err := results.Query()
+	if err != nil {
+		return fmt.Errorf("error querying results: %w", err)
+	}
+	for rows.Next() {
+		if err != nil {
+			err = rows.Err()
+		}
+		rows.Close()
+	}
+	return err
+}
+
+func (t *engine) Close() error {
+	if t.stopTelemetryRoutine != nil {
+		t.stopTelemetryRoutine()
+	}
+	if t.wgTelemetryRoutine != nil {
+		t.wgTelemetryRoutine.Wait()
+	}
+	return nil
+}
+
+func (t *engine) StartRoutineAsync() {
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	t.wgTelemetryRoutine = wg
+
+	ctx, stopTelemetryRoutine := context.WithCancel(context.Background())
+	t.stopTelemetryRoutine = stopTelemetryRoutine
+
+	go t.telemetrySync(ctx)
+}
+
+// telemetryStats are collected for telemetry information by all Promscale instances and sorted
+// in the instance information table.
+type telemetryStats struct {
+	samplesIngested                float64
+	promqlQueriesExecuted          float64
+	promqlQueriesTimedout          float64
+	promqlQueriesFailed            float64
+	traceQueriesExecuted           float64
+	traceDependencyQueriesExecuted float64
+}
+
+func (t *engine) telemetrySync(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			t.wgTelemetryRoutine.Done()
+			return
+		case <-time.After(time.Minute * 55):
+		}
+		t.metricMux.RLock()
```

Comment:
I don't see this lock being released if `len(t.metrics) != 0`

Reply:
The recent commits removes this part. Hence, resolving.